Fix decoding server's logprobs handling in Prefill/Decode disaggregation mode

Prowindy opened the pull request:
Signed-off-by: Cong Chen <prowindy@gmail.com>

This commit fixes issues with logprobs processing when using Prefill/Decode (PD) disaggregation, where decode servers receive invalid token IDs through the KV connector.

Changes:
- Filter out None values from prompt_logprobs for decode servers in completion API
- Skip processing prompt_logprobs on decode servers (identified by do_remote_decode=False)
- Store input kv_transfer_params in RequestState to identify decode servers
- Add proper handling for cases where all prompt logprobs are None

This prevents crashes when decode servers try to process invalid prompt token IDs.


## Purpose

## Test Plan

When using Prefill/Decode (PD) disaggregation with echo=True and logprobs enabled, the decode server crashes because it receives invalid prompt token IDs through the KV connector, causing an assertion failure when trying to process prompt_logprobs.

  Setup Requirements

  - Hardware: 8 GPUs (4 for prefill, 4 for decode)
  - Model: meta-llama/Llama-3.1-8B-Instruct
  - vLLM: Built with NIXL connector support
  - Dependencies: UCX libraries for NIXL

  ---
  Step 1: Start Prefill Server (4 GPUs)
```
  export VLLM_NIXL_SIDE_CHANNEL_HOST=0.0.0.0
  export VLLM_NIXL_SIDE_CHANNEL_PORT=8083
  export UCX_TLS=all
  export UCX_NET_DEVICES=all
  export VLLM_USE_V1=1

  CUDA_VISIBLE_DEVICES=0,1,2,3 vllm serve meta-llama/Llama-3.1-8B-Instruct \
      --host 0.0.0.0 \
      --port 8081 \
      --tensor-parallel-size 1 \
      --data-parallel-size 4 \
      --enforce-eager \
      --enable-prefix-caching \
      --gpu-memory-utilization 0.9 \
      --kv-transfer-config '{"kv_connector":"NixlConnector","kv_role":"kv_both","kv_connector_extra_config":{"backends":["UCX","GDS"],"http_port":8097}}' \
      --disable-log-stats
```
  ---
  Step 2: Start Decode Server (4 GPUs)
```
  export VLLM_NIXL_SIDE_CHANNEL_HOST=0.0.0.0
  export VLLM_NIXL_SIDE_CHANNEL_PORT=8093
  export UCX_TLS=all
  export UCX_NET_DEVICES=all
  export VLLM_USE_V1=1

  CUDA_VISIBLE_DEVICES=4,5,6,7 vllm serve meta-llama/Llama-3.1-8B-Instruct \
      --host 0.0.0.0 \
      --port 8082 \
      --tensor-parallel-size 1 \
      --data-parallel-size 4 \
      --compilation-config '{"cudagraph_mode":"FULL_DECODE_ONLY"}' \
      --kv-transfer-config '{"kv_connector":"NixlConnector","kv_role":"kv_both","kv_connector_extra_config":{"backends":["UCX","GDS"],"http_port":8098}}' \
      --disable-log-stats
```
  ---
  Step 3: Start PD Proxy Server
```
  cd /path/to/vllm
  python tests/v1/kv_connector/nixl_integration/toy_proxy_server.py \
      --host 127.0.0.1 \
      --port 8090 \
      --prefiller-hosts 127.0.0.1 \
      --prefiller-ports 8081 \
      --decoder-hosts 127.0.0.1 \
      --decoder-ports 8082
```
  ---
  Step 4: Reproduce the Bug

  Send a completion request with echo=True and logprobs enabled:
```
  curl http://localhost:8090/v1/completions \
    -H "Content-Type: application/json" \
    -d '{
      "model": "meta-llama/Llama-3.1-8B-Instruct",
      "prompt": "What is the capital of France?",
      "max_tokens": 10,
      "echo": true,
      "logprobs": 1
    }'
```
  Expected behavior without fix: Decode server crashes with:
(APIServer pid=2573880) ERROR 10-21 22:46:40 [v1/engine/async_llm.py:533]   File "/home/congc/gitrepos/vllm/vllm/v1/engine/logprobs.py", line 213, in update_from_output
(APIServer pid=2573880) ERROR 10-21 22:46:40 [v1/engine/async_llm.py:533]     self._update_prompt_logprobs(output.new_prompt_logprobs_tensors)

 
 

## Test Result

Step 5: Apply this patch and verify no more crashes

After applying this patch to vLLM, run the step 3 and 4 again. 
Expected behavior with fix: Request completes successfully.

Step 6: Verify the logprobs results

Run `toy_proxy_server_with_logprobs.py` instead of `toy_proxy_server.py`, because the new proxy supports concatenating prefill and decode logprobs.
```
  cd /path/to/vllm
  python tests/v1/kv_connector/nixl_integration/toy_proxy_server_with_logprobs.py \
      --host 127.0.0.1 \
      --port 8090 \
      --prefiller-hosts 127.0.0.1 \
      --prefiller-ports 8081 \
      --decoder-hosts 127.0.0.1 \
      --decoder-ports 8082
 ```
 
 Verification 1: send an identical request to prefill server, decode server and proxy respectively, their responses including logprobs are almost exactly same. 
 Verification 2: Run https://github.com/EleutherAI/lm-evaluation-harness and also bench test each of servers one by one. Their scores should be almost the same (about 33% for Llama 3.1)
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

Commit:
Fix logprobs handling in Prefill/Decode disaggregation mode

This commit fixes issues with logprobs processing when using Prefill/Decode
(PD) disaggregation, where decode servers receive invalid token IDs through
the KV connector.

Changes:
- Filter out None values from prompt_logprobs for decode servers in completion API
- Skip processing prompt_logprobs on decode servers (identified by do_remote_decode=False)
- Store input kv_transfer_params in RequestState to identify decode servers
- Add proper handling for cases where all prompt logprobs are None

This prevents crashes when decode servers try to process invalid prompt token IDs.

chaunceyjiang reviewed (commented):
Could you provide a minimal reproduction step?

Commit:
Fix prompt logprobs merging in PD disaggregation proxy

This commit fixes logprobs handling for Prefill/Decode (PD) disaggregation mode:

1. **serving_completion.py**: Preserve None values in prompt_logprobs instead of
   filtering them out. The None entry for BOS token is required to maintain
   alignment between prompt_logprobs and prompt_token_ids arrays.

2. **toy_proxy_server_with_logprobs.py**: Add new test proxy server that correctly
   merges logprobs from prefill and decode servers. The merge logic:
   - Extracts only prompt logprobs from prefill (excludes prefill's 1 output token)
   - Appends full decode response logprobs (both prompt and output)
   - Properly adjusts text_offset values for continuity
   - Handles token_logprobs, tokens, text_offset, and top_logprobs arrays

The proxy ensures that:
- Prompt logprobs [0:N] come from prefill (exact match)
- Output logprobs [N:M] come from decode (full generation)
- Array lengths match between proxy and standalone prefill server

njhill reviewed (changes requested):
Thanks for this @Prowindy!

I think the fix needs some changes. It should hopefully not require internal details of the NixlConnector to be added in the general code paths. I'll try to review properly soon.
